// just-bash-data/dist/index.cjs
'use strict';

var justBash = require('just-bash');
var jsDocStore = require('js-doc-store');
var jsVectorStore = require('js-vector-store');

// src/commands/db.ts

// src/lib/errors.ts
var EXIT = {
  OK: 0,
  RUNTIME: 1,
  USAGE: 2,
  NOT_FOUND: 3,
  AUTH: 4,
  VALIDATION: 5
};
var CommandError = class extends Error {
  exitCode;
  constructor(exitCode, message) {
    super(message);
    this.name = "CommandError";
    this.exitCode = exitCode;
  }
};
var COLL_NAME_RE = /^[a-zA-Z0-9_][a-zA-Z0-9_-]{0,63}$/;
var validateCollName = (name) => {
  if (!COLL_NAME_RE.test(name)) {
    throw new CommandError(
      EXIT.USAGE,
      `invalid collection name: '${name}' \u2014 allowed: ^[a-zA-Z0-9_][a-zA-Z0-9_-]{0,63}$ (no dots, slashes, or path components)`
    );
  }
};

// src/lib/args.ts
var DASH_DASH = "--";
var parseCommandArgs = (argv, spec = {}) => {
  const bool = new Set(spec.bool ?? []);
  const positional = [];
  const flags = /* @__PURE__ */ new Map();
  let passthrough = false;
  for (let i = 0; i < argv.length; i++) {
    const tok = argv[i];
    if (passthrough) {
      positional.push(tok);
      continue;
    }
    if (tok === DASH_DASH) {
      passthrough = true;
      continue;
    }
    if (tok.startsWith("--")) {
      const eq = tok.indexOf("=");
      if (eq !== -1) {
        flags.set(tok.slice(2, eq), tok.slice(eq + 1));
        continue;
      }
      const name = tok.slice(2);
      if (bool.has(name)) {
        flags.set(name, true);
        continue;
      }
      const next = argv[i + 1];
      if (next === void 0 || next.startsWith("--")) {
        throw new CommandError(EXIT.USAGE, `missing value for --${name}`);
      }
      flags.set(name, next);
      i++;
      continue;
    }
    positional.push(tok);
  }
  return { positional, flags };
};
var flagString = (flags, name) => {
  const v = flags.get(name);
  return typeof v === "string" ? v : void 0;
};
var flagBool = (flags, name) => flags.get(name) === true;

// src/lib/runner.ts
var runCommand = async (handler) => {
  try {
    const { stdout } = await handler();
    return { stdout, stderr: "", exitCode: EXIT.OK };
  } catch (err) {
    if (err instanceof CommandError) {
      return {
        stdout: "",
        stderr: `${err.message}
`,
        exitCode: err.exitCode
      };
    }
    const msg = err instanceof Error ? err.message : String(err);
    return { stdout: "", stderr: `${msg}
`, exitCode: EXIT.RUNTIME };
  }
};

// src/commands/db/shared.ts
var relaxJson = (input) => {
  const isIdStart = (c) => /[A-Za-z_$]/.test(c);
  const isIdCont = (c) => /[A-Za-z0-9_$]/.test(c);
  const isWs = (c) => c === " " || c === "	" || c === "\n" || c === "\r";
  let out = "";
  let i = 0;
  while (i < input.length) {
    const c = input[i];
    if (c === '"') {
      out += c;
      i++;
      while (i < input.length) {
        const ch = input[i];
        out += ch;
        if (ch === "\\" && i + 1 < input.length) {
          out += input[i + 1];
          i += 2;
          continue;
        }
        i++;
        if (ch === '"') break;
      }
      continue;
    }
    if (c === "'") {
      out += '"';
      i++;
      while (i < input.length && input[i] !== "'") {
        const ch = input[i];
        if (ch === "\\" && i + 1 < input.length) {
          const next = input[i + 1];
          if (next === "'") {
            out += "'";
          } else {
            out += "\\" + next;
          }
          i += 2;
          continue;
        }
        if (ch === '"') {
          out += '\\"';
          i++;
          continue;
        }
        out += ch;
        i++;
      }
      out += '"';
      i++;
      continue;
    }
    if (isIdStart(c)) {
      const start = i;
      while (i < input.length && isIdCont(input[i])) i++;
      const word = input.slice(start, i);
      let j = i;
      while (j < input.length && isWs(input[j])) j++;
      if (input[j] === ":" && word !== "true" && word !== "false" && word !== "null") {
        out += `"${word}"`;
      } else {
        out += word;
      }
      continue;
    }
    if (c === ",") {
      let j = i + 1;
      while (j < input.length && isWs(input[j])) j++;
      const next = input[j];
      if (next === "}" || next === "]") {
        i++;
        continue;
      }
    }
    out += c;
    i++;
  }
  return out;
};
var FILTER_OPERATOR_NAMES = /* @__PURE__ */ new Set([
  "eq",
  "ne",
  "gt",
  "gte",
  "lt",
  "lte",
  "in",
  "nin",
  "exists",
  "regex",
  "contains",
  "size",
  "and",
  "or",
  "not"
]);
var validateFilterOperators = (value, fieldName, path = "") => {
  if (Array.isArray(value)) {
    for (let i = 0; i < value.length; i++) {
      validateFilterOperators(value[i], fieldName, `${path}[${i}]`);
    }
    return;
  }
  if (typeof value !== "object" || value === null) return;
  for (const [k, v] of Object.entries(value)) {
    if (!k.startsWith("$") && FILTER_OPERATOR_NAMES.has(k)) {
      const here = path ? `${path}.${k}` : k;
      throw new CommandError(
        EXIT.VALIDATION,
        `validation: ${fieldName} operator '${k}' at ${here} is missing $ prefix \u2014 did you mean '$${k}'?`
      );
    }
    validateFilterOperators(v, fieldName, path ? `${path}.${k}` : k);
  }
};
var PIPELINE_STAGE_NAMES = /* @__PURE__ */ new Set([
  "match",
  "lookup",
  "group",
  "sort",
  "limit",
  "skip",
  "project",
  "unwind"
]);
var GROUP_ACCUMULATOR_NAMES = /* @__PURE__ */ new Set([
  "count",
  "sum",
  "avg",
  "min",
  "max",
  "push",
  "first",
  "last"
]);
var UPDATE_OPERATOR_NAMES = /* @__PURE__ */ new Set([
  "set",
  "unset",
  "inc",
  "push",
  "pull",
  "rename"
]);
var validatePipeline = (pipeline, fieldName) => {
  if (!Array.isArray(pipeline)) return;
  for (let i = 0; i < pipeline.length; i++) {
    const stage = pipeline[i];
    if (typeof stage !== "object" || stage === null || Array.isArray(stage)) continue;
    const stageObj = stage;
    for (const [stageKey, stageVal] of Object.entries(stageObj)) {
      if (!stageKey.startsWith("$") && PIPELINE_STAGE_NAMES.has(stageKey)) {
        throw new CommandError(
          EXIT.VALIDATION,
          `validation: ${fieldName} stage '${stageKey}' at [${i}].${stageKey} is missing $ prefix \u2014 did you mean '$${stageKey}'?`
        );
      }
      if (stageKey === "$match") {
        validateFilterOperators(stageVal, `${fieldName}[${i}].$match`);
      }
      if (stageKey === "$group" && stageVal !== null && typeof stageVal === "object" && !Array.isArray(stageVal)) {
        for (const [groupKey, groupVal] of Object.entries(stageVal)) {
          if (groupKey === "_id") continue;
          if (groupVal === null || typeof groupVal !== "object" || Array.isArray(groupVal)) continue;
          for (const accKey of Object.keys(groupVal)) {
            if (!accKey.startsWith("$") && GROUP_ACCUMULATOR_NAMES.has(accKey)) {
              throw new CommandError(
                EXIT.VALIDATION,
                `validation: ${fieldName} accumulator '${accKey}' at [${i}].$group.${groupKey}.${accKey} is missing $ prefix \u2014 did you mean '$${accKey}'?`
              );
            }
          }
        }
      }
    }
  }
};
var validateUpdateOperators = (doc, fieldName) => {
  if (doc === null || typeof doc !== "object" || Array.isArray(doc)) return;
  for (const k of Object.keys(doc)) {
    if (!k.startsWith("$") && UPDATE_OPERATOR_NAMES.has(k)) {
      throw new CommandError(
        EXIT.VALIDATION,
        `validation: ${fieldName} operator '${k}' is missing $ prefix \u2014 did you mean '$${k}'?`
      );
    }
  }
};
var parseJson = (arg, ctx, fieldName, emptyPolicy = "reject", validateFilter = false) => {
  if (arg === void 0) {
    throw new CommandError(EXIT.USAGE, `missing ${fieldName} argument`);
  }
  const text = arg === "-" ? ctx.stdin : arg;
  if (text.trim() === "") {
    if (emptyPolicy === "filter") return {};
    if (emptyPolicy === "pipeline") return [];
    throw new CommandError(
      EXIT.USAGE,
      `${fieldName} cannot be empty (use '{}' explicitly if you mean an empty object)`
    );
  }
  let value;
  try {
    value = JSON.parse(text);
  } catch {
    try {
      value = JSON.parse(relaxJson(text));
    } catch {
      throw new CommandError(EXIT.USAGE, `invalid json: ${fieldName}`);
    }
  }
  if (validateFilter) {
    validateFilterOperators(value, fieldName);
  }
  return value;
};
var ensureSingleStdinDash = (positionals) => {
  const dashCount = positionals.filter((p) => p === "-").length;
  if (dashCount > 1) {
    throw new CommandError(
      EXIT.USAGE,
      "usage: only one positional may read from stdin"
    );
  }
};
var resolveToken = (ctx, parsed) => {
  return flagString(parsed.flags, "token") ?? ctx.env.get("AUTH_TOKEN");
};
var requireAuth = async (reg, ctx, parsed) => {
  if (!reg.opts.authSecret) {
    return { sub: "anonymous", roles: [] };
  }
  const token = resolveToken(ctx, parsed);
  if (!token) {
    throw new CommandError(EXIT.AUTH, "auth: missing token");
  }
  const auth = await reg.getAuth();
  const payload = await auth.verify(token);
  if (!payload) {
    throw new CommandError(EXIT.AUTH, "auth: invalid token");
  }
  if (typeof payload.exp === "number" && Date.now() / 1e3 > payload.exp) {
    throw new CommandError(EXIT.AUTH, "auth: expired token");
  }
  return { sub: payload.sub, roles: payload.roles };
};
var requireRole = async (reg, ctx, parsed, role) => {
  const payload = await requireAuth(reg, ctx, parsed);
  if (reg.opts.authSecret && !payload.roles.includes(role)) {
    throw new CommandError(EXIT.AUTH, `auth: role required: ${role}`);
  }
  return payload;
};
var ensureCollExists = (reg, coll) => {
  const meta = reg.mem.readJson(`${coll}.meta.json`);
  const docs = reg.mem.readJson(`${coll}.docs.json`);
  if (meta === null && docs === null) {
    throw new CommandError(EXIT.NOT_FOUND, `not found: ${coll}`);
  }
};
var isFilter = (v) => typeof v === "object" && v !== null && !Array.isArray(v);

// src/commands/db/auth.ts
var requireAuthSecret = (reg) => {
  if (!reg.opts.authSecret) {
    throw new CommandError(
      EXIT.AUTH,
      "auth: not configured (PluginOptions.authSecret missing)"
    );
  }
};
var authHandler = async (reg, ctx, parsed) => {
  const sub = parsed.positional[1];
  switch (sub) {
    case "register":
      return registerSub(reg, ctx, parsed);
    case "login":
      return loginSub(reg, parsed);
    case "verify":
      return verifySub(reg, ctx, parsed);
    case "logout":
      return logoutSub(reg, ctx, parsed);
    case "role":
      return roleSub(reg, ctx, parsed);
    default:
      throw new CommandError(
        EXIT.USAGE,
        "usage: db auth <register|login|verify|logout|role> [...]"
      );
  }
};
var registerSub = async (reg, _ctx, parsed) => {
  requireAuthSecret(reg);
  const user = parsed.positional[2];
  const pass = parsed.positional[3];
  if (!user || !pass) {
    throw new CommandError(
      EXIT.USAGE,
      "usage: db auth register <user> <pass> [--roles=a,b]"
    );
  }
  const auth = await reg.getAuth();
  const rolesFlag = parsed.flags.get("roles");
  const profile = {};
  if (typeof rolesFlag === "string") {
    profile["roles"] = rolesFlag.split(",").map((s) => s.trim()).filter(Boolean);
  }
  try {
    const created = await auth.register(user, pass, profile);
    return {
      stdout: JSON.stringify({ user: created["email"], id: created._id })
    };
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    if (msg.includes("Unique constraint") || msg.toLowerCase().includes("duplicate")) {
      throw new CommandError(EXIT.VALIDATION, `validation: duplicate user: ${user}`);
    }
    if (msg.includes("Password must be") || msg.includes("Email and password")) {
      throw new CommandError(EXIT.USAGE, msg);
    }
    throw err;
  }
};
var loginSub = async (reg, parsed) => {
  requireAuthSecret(reg);
  const user = parsed.positional[2];
  const pass = parsed.positional[3];
  if (!user || !pass) {
    throw new CommandError(EXIT.USAGE, "usage: db auth login <user> <pass>");
  }
  const auth = await reg.getAuth();
  try {
    const result = await auth.login(user, pass);
    return {
      stdout: JSON.stringify({
        token: result.token,
        expiresAt: new Date(
          Date.now() + 86400 * 1e3
        ).toISOString()
      })
    };
  } catch {
    throw new CommandError(EXIT.AUTH, "auth: invalid credentials");
  }
};
var verifySub = async (reg, ctx, parsed) => {
  requireAuthSecret(reg);
  const token = resolveToken(ctx, parsed);
  if (!token) throw new CommandError(EXIT.AUTH, "auth: missing token");
  const auth = await reg.getAuth();
  const payload = await auth.verify(token);
  if (!payload) throw new CommandError(EXIT.AUTH, "auth: invalid token");
  return {
    stdout: JSON.stringify({
      user: payload.email,
      roles: payload.roles,
      expiresAt: new Date(payload.exp * 1e3).toISOString()
    })
  };
};
var logoutSub = async (reg, ctx, parsed) => {
  requireAuthSecret(reg);
  const token = resolveToken(ctx, parsed);
  if (!token) throw new CommandError(EXIT.AUTH, "auth: missing token");
  const auth = await reg.getAuth();
  const payload = await auth.verify(token);
  if (!payload) throw new CommandError(EXIT.AUTH, "auth: invalid token");
  if (flagBool(parsed.flags, "all")) {
    auth.logoutAll(payload.sub);
  } else {
    auth.logout(token);
  }
  return { stdout: JSON.stringify({ loggedOut: true }) };
};
var roleSub = async (reg, ctx, parsed) => {
  requireAuthSecret(reg);
  await requireRole(reg, ctx, parsed, "admin");
  const op = parsed.positional[2];
  const targetUser = parsed.positional[3];
  const role = parsed.positional[4];
  if (!op || !targetUser || !role) {
    throw new CommandError(
      EXIT.USAGE,
      "usage: db auth role <assign|remove> <user-id> <role>"
    );
  }
  const auth = await reg.getAuth();
  try {
    if (op === "assign") await auth.assignRole(targetUser, role);
    else if (op === "remove") await auth.removeRole(targetUser, role);
    else throw new CommandError(EXIT.USAGE, `unknown role op: ${op}`);
  } catch (err) {
    if (err instanceof CommandError) throw err;
    const msg = err instanceof Error ? err.message : String(err);
    if (msg.includes("User not found")) {
      throw new CommandError(EXIT.NOT_FOUND, `not found: user ${targetUser}`);
    }
    throw err;
  }
  return { stdout: JSON.stringify({ user: targetUser, op }) };
};

// src/commands/db/crud.ts
var insertHandler = async (reg, ctx, parsed, coll) => {
  await requireAuth(reg, ctx, parsed);
  const docArg = parsed.positional[2];
  const doc = parseJson(docArg, ctx, "document", "reject");
  if (!isFilter(doc)) {
    throw new CommandError(EXIT.USAGE, "document must be an object");
  }
  try {
    const result = reg.getDocStore().collection(coll).insert(doc);
    return { stdout: JSON.stringify({ _id: result._id }) };
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    if (msg.includes("Unique constraint")) {
      throw new CommandError(EXIT.VALIDATION, `validation: ${msg}`);
    }
    throw err;
  }
};
var buildCursor = (reg, coll, filter, parsed, options = null) => {
  let cursor = reg.getDocStore().collection(coll).find(filter);
  const sortFlag = flagString(parsed.flags, "sort");
  if (sortFlag) {
    const [field, dir] = sortFlag.split(":");
    if (!field) throw new CommandError(EXIT.USAGE, "invalid --sort");
    const order = dir === "-1" ? -1 : 1;
    cursor = cursor.sort({ [field]: order });
  } else if (options && isFilter(options["sort"])) {
    cursor = cursor.sort(options["sort"]);
  }
  const skipFlag = flagString(parsed.flags, "skip");
  const skipN = skipFlag !== void 0 ? Number(skipFlag) : typeof options?.["skip"] === "number" ? options["skip"] : void 0;
  if (skipN !== void 0) cursor = cursor.skip(skipN);
  const limitFlag = flagString(parsed.flags, "limit");
  const limitN = limitFlag !== void 0 ? Number(limitFlag) : typeof options?.["limit"] === "number" ? options["limit"] : void 0;
  if (limitN !== void 0) cursor = cursor.limit(limitN);
  const projectFlag = flagString(parsed.flags, "project");
  if (projectFlag) {
    const spec = {};
    for (const f of projectFlag.split(",")) spec[f.trim()] = 1;
    cursor = cursor.project(spec);
  } else if (options && isFilter(options["project"])) {
    cursor = cursor.project(options["project"]);
  }
  return cursor.toArray();
};
var findHandler = async (reg, ctx, parsed, coll) => {
  ensureCollExists(reg, coll);
  const filter = parseJson(parsed.positional[2] ?? "{}", ctx, "filter", "filter", true);
  if (!isFilter(filter)) {
    throw new CommandError(EXIT.USAGE, "filter must be an object");
  }
  let options = null;
  const optionsArg = parsed.positional[3];
  if (optionsArg !== void 0) {
    const parsedOpts = parseJson(optionsArg, ctx, "options", "reject");
    if (isFilter(parsedOpts)) options = parsedOpts;
  }
  const docs = buildCursor(reg, coll, filter, parsed, options);
  return { stdout: JSON.stringify(docs) };
};
var countHandler = async (reg, _ctx, parsed, coll) => {
  ensureCollExists(reg, coll);
  const filter = parseJson(parsed.positional[2] ?? "{}", _ctx, "filter", "filter", true);
  if (!isFilter(filter)) {
    throw new CommandError(EXIT.USAGE, "filter must be an object");
  }
  const n = reg.getDocStore().collection(coll).count(filter);
  return { stdout: JSON.stringify({ count: n }) };
};
var updateHandler = async (reg, ctx, parsed, coll) => {
  await requireAuth(reg, ctx, parsed);
  ensureCollExists(reg, coll);
  const filter = parseJson(parsed.positional[2], ctx, "filter", "reject", true);
  const update = parseJson(parsed.positional[3], ctx, "update", "reject");
  validateUpdateOperators(update, "update");
  if (!isFilter(filter) || !isFilter(update)) {
    throw new CommandError(EXIT.USAGE, "filter and update must be objects");
  }
  const c = reg.getDocStore().collection(coll);
  const many = flagBool(parsed.flags, "many");
  try {
    const before = c.count(filter);
    const modified = many ? c.updateMany(filter, update) : c.update(filter, update);
    const matched = many ? before : before > 0 ? 1 : 0;
    return {
      stdout: JSON.stringify({ matched, modified })
    };
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    if (msg.includes("Unique constraint")) {
      throw new CommandError(EXIT.VALIDATION, `validation: ${msg}`);
    }
    throw err;
  }
};
var removeHandler = async (reg, ctx, parsed, coll) => {
  await requireAuth(reg, ctx, parsed);
  ensureCollExists(reg, coll);
  const filter = parseJson(parsed.positional[2], ctx, "filter", "reject", true);
  if (!isFilter(filter)) {
    throw new CommandError(EXIT.USAGE, "filter must be an object");
  }
  const c = reg.getDocStore().collection(coll);
  const many = flagBool(parsed.flags, "many");
  const removed = many ? c.removeMany(filter) : c.remove(filter);
  return { stdout: JSON.stringify({ removed }) };
};
var aggregateHandler = async (reg, ctx, parsed, coll) => {
  ensureCollExists(reg, coll);
  const pipeline = parseJson(parsed.positional[2], ctx, "pipeline", "pipeline");
  if (!Array.isArray(pipeline)) {
    throw new CommandError(EXIT.USAGE, "pipeline must be an array");
  }
  validatePipeline(pipeline, "pipeline");
  let pipe = reg.getDocStore().collection(coll).aggregate();
  for (const stage of pipeline) {
    if (!isFilter(stage)) {
      throw new CommandError(EXIT.USAGE, "each pipeline stage must be an object");
    }
    const keys = Object.keys(stage);
    if (keys.length !== 1) {
      throw new CommandError(EXIT.USAGE, "each stage must have exactly one operator");
    }
    const op = keys[0];
    const arg = stage[op];
    switch (op) {
      case "$match":
        pipe = pipe.match(arg);
        break;
      case "$lookup":
        pipe = pipe.lookup(arg);
        break;
      case "$group": {
        if (!isFilter(arg)) {
          throw new CommandError(EXIT.USAGE, "$group must be an object");
        }
        const spec = arg;
        const idVal = spec["_id"];
        let field = null;
        if (typeof idVal === "string") {
          field = idVal.startsWith("$") ? idVal.slice(1) : idVal;
        }
        const accumulators = {};
        for (const [accName, accDef] of Object.entries(spec)) {
          if (accName === "_id") continue;
          if (!isFilter(accDef)) continue;
          const out = {};
          for (const [opName, opVal] of Object.entries(accDef)) {
            if (opName === "$sum" && typeof opVal !== "string") {
              out["$count"] = 1;
              continue;
            }
            out[opName] = typeof opVal === "string" && opVal.startsWith("$") ? opVal.slice(1) : opVal;
          }
          accumulators[accName] = out;
        }
        pipe = pipe.group(field, accumulators);
        break;
      }
      case "$sort":
        pipe = pipe.sort(arg);
        break;
      case "$limit":
        pipe = pipe.limit(arg);
        break;
      case "$skip":
        pipe = pipe.skip(arg);
        break;
      case "$project":
        pipe = pipe.project(arg);
        break;
      case "$unwind": {
        const unwindField = typeof arg === "string" && arg.startsWith("$") ? arg.slice(1) : arg;
        pipe = pipe.unwind(unwindField);
        break;
      }
      default:
        throw new CommandError(EXIT.USAGE, `unknown aggregation operator: ${op}`);
    }
  }
  return { stdout: JSON.stringify(pipe.toArray()) };
};

// src/commands/db/meta.ts
var indexHandler = async (reg, ctx, parsed, coll) => {
  const sub = parsed.positional[2];
  const c = reg.getDocStore().collection(coll);
  switch (sub) {
    case "create": {
      await requireAuth(reg, ctx, parsed);
      const field = parsed.positional[3];
      if (!field) throw new CommandError(EXIT.USAGE, "usage: db <coll> index create <field>");
      const sorted = flagBool(parsed.flags, "sorted");
      const unique = flagBool(parsed.flags, "unique");
      try {
        c.createIndex(field, { type: sorted ? "sorted" : "hash", unique });
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        if (msg.includes("Unique constraint")) {
          throw new CommandError(EXIT.VALIDATION, `validation: ${msg}`);
        }
        throw err;
      }
      return {
        stdout: JSON.stringify({ created: field, type: sorted ? "sorted" : "hash" })
      };
    }
    case "drop": {
      await requireAuth(reg, ctx, parsed);
      const field = parsed.positional[3];
      if (!field) throw new CommandError(EXIT.USAGE, "usage: db <coll> index drop <field>");
      ensureCollExists(reg, coll);
      const exists = c.getIndexes().some((idx) => idx.field === field);
      if (!exists) {
        throw new CommandError(EXIT.NOT_FOUND, `not found: index ${field}`);
      }
      c.dropIndex(field);
      return { stdout: JSON.stringify({ dropped: field }) };
    }
    case "list": {
      ensureCollExists(reg, coll);
      return { stdout: JSON.stringify(c.getIndexes()) };
    }
    default:
      throw new CommandError(
        EXIT.USAGE,
        "usage: db <coll> index <create|drop|list> [...]"
      );
  }
};
var dropHandler = async (reg, ctx, parsed, coll) => {
  await requireRole(reg, ctx, parsed, "admin");
  ensureCollExists(reg, coll);
  reg.getDocStore().drop(coll);
  return { stdout: JSON.stringify({ dropped: coll }) };
};
var statsHandler = async (reg, _ctx, _parsed, coll) => {
  ensureCollExists(reg, coll);
  const c = reg.getDocStore().collection(coll);
  const docs = reg.mem.readJson(`${coll}.docs.json`);
  const sizeBytes = docs ? new TextEncoder().encode(JSON.stringify(docs)).byteLength : 0;
  return {
    stdout: JSON.stringify({
      count: c.count(),
      indexes: c.getIndexes(),
      sizeBytes
    })
  };
};
var exportHandler = async (reg, _ctx, _parsed, coll) => {
  ensureCollExists(reg, coll);
  const docs = reg.getDocStore().collection(coll).export();
  return { stdout: JSON.stringify({ exported: docs.length, docs }) };
};
var importHandler = async (reg, ctx, parsed, coll) => {
  await requireAuth(reg, ctx, parsed);
  const arg = parsed.positional[2];
  const data = parseJson(arg, ctx, "import data", "reject");
  if (!Array.isArray(data)) {
    throw new CommandError(EXIT.USAGE, "import expects an array of documents");
  }
  for (let i = 0; i < data.length; i++) {
    if (!isFilter(data[i])) {
      throw new CommandError(
        EXIT.USAGE,
        `import item at index ${i} is not an object`
      );
    }
  }
  reg.getDocStore().collection(coll).import(data);
  return { stdout: JSON.stringify({ imported: data.length }) };
};

// src/commands/db.ts
var FLAG_SPEC = {
  bool: ["many", "json", "sorted", "unique", "all"],
  string: ["token", "sort", "limit", "skip", "project", "roles"]
};
var buildDbCommand = (provide) => justBash.defineCommand(
  "db",
  async (args, ctx) => runCommand(async () => {
    const reg = provide(ctx.fs);
    const parsed = parseCommandArgs(args, FLAG_SPEC);
    ensureSingleStdinDash(parsed.positional);
    const first = parsed.positional[0];
    if (!first) {
      throw new CommandError(
        EXIT.USAGE,
        "usage: db <coll|auth> <subcommand> [args...]"
      );
    }
    await reg.ensureHydrated();
    let result;
    let mutated = false;
    if (first === "auth") {
      result = await authHandler(reg, ctx, parsed);
      const sub = parsed.positional[1];
      mutated = sub === "register" || sub === "login" || sub === "logout" || sub === "role";
    } else {
      const coll = first;
      validateCollName(coll);
      const sub = parsed.positional[1];
      if (!sub) {
        throw new CommandError(
          EXIT.USAGE,
          "usage: db <coll> <subcommand> [args...]"
        );
      }
      switch (sub) {
        case "insert":
          result = await insertHandler(reg, ctx, parsed, coll);
          mutated = true;
          break;
        case "find":
          result = await findHandler(reg, ctx, parsed, coll);
          break;
        case "count":
          result = await countHandler(reg, ctx, parsed, coll);
          break;
        case "update":
          result = await updateHandler(reg, ctx, parsed, coll);
          mutated = true;
          break;
        case "remove":
          result = await removeHandler(reg, ctx, parsed, coll);
          mutated = true;
          break;
        case "aggregate":
          result = await aggregateHandler(reg, ctx, parsed, coll);
          break;
        case "index": {
          result = await indexHandler(reg, ctx, parsed, coll);
          const idxOp = parsed.positional[2];
          mutated = idxOp === "create" || idxOp === "drop";
          break;
        }
        case "drop":
          result = await dropHandler(reg, ctx, parsed, coll);
          mutated = true;
          break;
        case "stats":
          result = await statsHandler(reg, ctx, parsed, coll);
          break;
        case "export":
          result = await exportHandler(reg, ctx, parsed, coll);
          break;
        case "import":
          result = await importHandler(reg, ctx, parsed, coll);
          mutated = true;
          break;
        default:
          throw new CommandError(EXIT.USAGE, `unknown subcommand: db <coll> ${sub}`);
      }
    }
    if (mutated) {
      reg.getDocStore().flush();
      await reg.flushIfDirty();
    }
    return result;
  })
);

// src/commands/vec/shared.ts
var QUANTIZE_VALUES = ["float32", "int8", "polar", "binary"];
var METRIC_VALUES = ["cosine", "euclidean", "dot", "manhattan"];
var parseVectorJson = (arg, ctx) => {
  if (arg === void 0) {
    throw new CommandError(EXIT.USAGE, "missing vector argument");
  }
  const text = arg === "-" ? ctx.stdin : arg;
  let parsed;
  try {
    parsed = JSON.parse(text);
  } catch {
    throw new CommandError(EXIT.USAGE, "invalid vector: not valid JSON");
  }
  if (!Array.isArray(parsed) || parsed.some((n) => typeof n !== "number" || !Number.isFinite(n))) {
    throw new CommandError(EXIT.USAGE, "invalid vector: expected number[]");
  }
  return parsed;
};
var requireVecColl = (reg, coll) => {
  validateCollName(coll);
  const entry = reg.getVectorCollection(coll);
  if (!entry) throw new CommandError(EXIT.NOT_FOUND, `not found: ${coll}`);
  return entry;
};
var validateQuantize = (v) => {
  if (QUANTIZE_VALUES.includes(v)) return v;
  throw new CommandError(
    EXIT.USAGE,
    `invalid --quantize: ${v} (allowed: ${QUANTIZE_VALUES.join("|")})`
  );
};
var validateMetric = (v) => {
  if (METRIC_VALUES.includes(v)) return v;
  throw new CommandError(
    EXIT.USAGE,
    `invalid --metric: ${v} (allowed: ${METRIC_VALUES.join("|")})`
  );
};
var checkVectorRecord = (rec, dim) => {
  if (!rec || typeof rec !== "object" || Array.isArray(rec)) {
    return { ok: false, reason: "not an object" };
  }
  const r = rec;
  const id = r["id"];
  const vec = r["vector"];
  if (typeof id !== "string") return { ok: false, reason: "missing or non-string id" };
  if (!Array.isArray(vec)) return { ok: false, reason: "missing or non-array vector" };
  if (vec.some((n) => typeof n !== "number" || !Number.isFinite(n))) {
    return { ok: false, reason: "non-finite numbers in vector" };
  }
  if (vec.length !== dim) {
    return { ok: false, reason: `dim mismatch (${vec.length} vs ${dim})` };
  }
  return { ok: true, id, vector: vec };
};

// src/commands/vec/ops.ts
var createOp = async (reg, _ctx, parsed) => {
  const coll = parsed.positional[1];
  if (!coll) throw new CommandError(EXIT.USAGE, "usage: vec create <coll> --dim N");
  validateCollName(coll);
  if (reg.getVectorCollection(coll)) {
    throw new CommandError(EXIT.VALIDATION, `validation: collection exists: ${coll}`);
  }
  const dimStr = flagString(parsed.flags, "dim");
  if (!dimStr) throw new CommandError(EXIT.USAGE, "missing --dim");
  const dim = Number(dimStr);
  if (!Number.isInteger(dim) || dim <= 0) {
    throw new CommandError(EXIT.USAGE, `invalid --dim: ${dimStr}`);
  }
  if (dim > 65536) {
    throw new CommandError(EXIT.USAGE, `--dim too large: ${dim} (max 65536)`);
  }
  const quantizeFlag = flagString(parsed.flags, "quantize");
  const quantize = quantizeFlag ? validateQuantize(quantizeFlag) : "float32";
  const metricFlag = flagString(parsed.flags, "metric");
  const metric = metricFlag ? validateMetric(metricFlag) : "cosine";
  const ivfClustersStr = flagString(parsed.flags, "ivf-clusters");
  const ivfProbesStr = flagString(parsed.flags, "ivf-probes");
  const ivfRequested = flagBool(parsed.flags, "ivf") || ivfClustersStr !== void 0 || ivfProbesStr !== void 0;
  let ivf;
  if (ivfRequested) {
    const numClusters = ivfClustersStr ? Number(ivfClustersStr) : 100;
    const numProbes = ivfProbesStr ? Number(ivfProbesStr) : 10;
    if (!Number.isInteger(numClusters) || numClusters <= 0) {
      throw new CommandError(EXIT.USAGE, `invalid --ivf-clusters: ${ivfClustersStr}`);
    }
    if (!Number.isInteger(numProbes) || numProbes <= 0) {
      throw new CommandError(EXIT.USAGE, `invalid --ivf-probes: ${ivfProbesStr}`);
    }
    if (numProbes > numClusters) {
      throw new CommandError(
        EXIT.USAGE,
        `--ivf-probes (${numProbes}) cannot exceed --ivf-clusters (${numClusters})`
      );
    }
    ivf = { numClusters, numProbes };
  }
  const entry = reg.registerVectorCollection(coll, dim, quantize, metric, ivf);
  reg.persistVectorRegistry();
  const out = {
    coll,
    dim: entry.dim,
    quantize: entry.quantize,
    metric: entry.metric
  };
  if (entry.ivf) out["ivf"] = entry.ivf;
  return { stdout: JSON.stringify(out) };
};
var ivfOp = async (reg, _ctx, parsed) => {
  const sub = parsed.positional[1];
  const coll = parsed.positional[2];
  if (!coll) {
    throw new CommandError(EXIT.USAGE, "usage: vec ivf <build|stats|drop> <coll>");
  }
  const entry = requireVecColl(reg, coll);
  if (!entry.ivfIndex) {
    throw new CommandError(
      EXIT.USAGE,
      `collection '${coll}' was not created with --ivf; recreate with --ivf or --ivf-clusters`
    );
  }
  switch (sub) {
    case "build": {
      const sampleDimsStr = flagString(parsed.flags, "sample-dims");
      const sampleDims = sampleDimsStr ? Number(sampleDimsStr) : Math.min(128, entry.dim);
      try {
        const result = entry.ivfIndex.build(coll, sampleDims);
        return { stdout: JSON.stringify({ coll, ...result }) };
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        if (msg.includes("vac\xEDa") || msg.toLowerCase().includes("empty")) {
          throw new CommandError(EXIT.VALIDATION, `validation: collection is empty: ${coll}`);
        }
        throw err;
      }
    }
    case "stats": {
      const stats = entry.ivfIndex.indexStats(coll);
      if (!stats) {
        throw new CommandError(
          EXIT.NOT_FOUND,
          `not found: ivf index not built for ${coll} (run: vec ivf build ${coll})`
        );
      }
      return {
        stdout: JSON.stringify({
          coll,
          ...stats,
          numVectors: entry.store.count(coll)
        })
      };
    }
    case "drop": {
      if (!entry.ivfIndex.hasIndex(coll)) {
        throw new CommandError(EXIT.NOT_FOUND, `not found: ivf index for ${coll}`);
      }
      entry.ivfIndex.dropIndex(coll);
      return { stdout: JSON.stringify({ dropped: coll }) };
    }
    default:
      throw new CommandError(
        EXIT.USAGE,
        `usage: vec ivf <build|stats|drop> <coll>`
      );
  }
};
var storeOp = async (reg, ctx, parsed) => {
  const coll = parsed.positional[1];
  const id = parsed.positional[2];
  const vecArg = parsed.positional[3];
  if (!coll || !id) {
    throw new CommandError(EXIT.USAGE, "usage: vec store <coll> <id> <vector>");
  }
  const entry = requireVecColl(reg, coll);
  const vector = parseVectorJson(vecArg, ctx);
  if (vector.length !== entry.dim) {
    throw new CommandError(
      EXIT.VALIDATION,
      `validation: dim mismatch (got ${vector.length}, expected ${entry.dim})`
    );
  }
  const metaFlag = flagString(parsed.flags, "meta");
  let meta = {};
  if (metaFlag !== void 0) {
    try {
      const parsedMeta = JSON.parse(metaFlag);
      if (typeof parsedMeta === "object" && parsedMeta !== null && !Array.isArray(parsedMeta)) {
        meta = parsedMeta;
      }
    } catch {
      throw new CommandError(EXIT.USAGE, "invalid --meta: not valid JSON");
    }
  }
  entry.store.set(coll, id, vector, meta);
  return { stdout: JSON.stringify({ id }) };
};
var storeBatchOp = async (reg, ctx, parsed) => {
  const coll = parsed.positional[1];
  const src = parsed.positional[2];
  if (!coll || !src) {
    throw new CommandError(EXIT.USAGE, "usage: vec store-batch <coll> <jsonl-path-or-->");
  }
  const entry = requireVecColl(reg, coll);
  let text;
  if (src === "-") {
    text = ctx.stdin;
  } else {
    try {
      text = await ctx.fs.readFile(src, "utf8");
    } catch {
      throw new CommandError(EXIT.USAGE, `cannot read jsonl: ${src}`);
    }
  }
  const lines = text.split(/\r?\n/).filter((l) => l.length > 0);
  let stored = 0;
  let skipped = 0;
  const errors = [];
  const skip = (line, reason) => {
    skipped++;
    if (errors.length < 20) errors.push({ line, reason });
  };
  for (let i = 0; i < lines.length; i++) {
    const lineText = lines[i];
    let parsedRec;
    try {
      parsedRec = JSON.parse(lineText);
    } catch {
      skip(i + 1, "invalid json");
      continue;
    }
    const check = checkVectorRecord(parsedRec, entry.dim);
    if (!check.ok) {
      skip(i + 1, check.reason);
      continue;
    }
    if (entry.store.has(coll, check.id)) {
      throw new CommandError(EXIT.VALIDATION, `validation: id collision: ${check.id}`);
    }
    const rec = parsedRec;
    const recMeta = rec["metadata"] ?? rec["meta"];
    const meta = recMeta && typeof recMeta === "object" && !Array.isArray(recMeta) ? recMeta : {};
    entry.store.set(coll, check.id, check.vector, meta);
    stored++;
  }
  return { stdout: JSON.stringify({ stored, skipped, errors }) };
};
var searchOp = async (reg, ctx, parsed) => {
  const coll = parsed.positional[1];
  const vecArg = parsed.positional[2];
  if (!coll) throw new CommandError(EXIT.USAGE, "usage: vec search <coll> <vector>");
  const entry = requireVecColl(reg, coll);
  const query = parseVectorJson(vecArg, ctx);
  const k = Number(flagString(parsed.flags, "k") ?? "10");
  const metricOverride = flagString(parsed.flags, "metric");
  const metric = metricOverride ? validateMetric(metricOverride) : entry.metric;
  const matryoshka = flagString(parsed.flags, "matryoshka");
  const noIvf = flagBool(parsed.flags, "no-ivf");
  const useIvf = !noIvf && !metricOverride && entry.ivfIndex !== void 0 && entry.ivfIndex.hasIndex(coll);
  let hits;
  if (useIvf) {
    hits = matryoshka ? entry.ivfIndex.matryoshkaSearch(
      coll,
      query,
      k,
      matryoshka.split(",").map((s) => Number(s.trim()))
    ) : entry.ivfIndex.search(coll, query, k);
  } else {
    hits = matryoshka ? entry.store.matryoshkaSearch(
      coll,
      query,
      k,
      matryoshka.split(",").map((s) => Number(s.trim())),
      metric
    ) : entry.store.search(coll, query, k, 0, metric, null);
  }
  return { stdout: JSON.stringify(hits) };
};
var searchAcrossOp = async (reg, ctx, parsed) => {
  const collsArg = parsed.positional[1];
  const vecArg = parsed.positional[2];
  if (!collsArg) {
    throw new CommandError(EXIT.USAGE, "usage: vec search-across <coll-csv> <vector>");
  }
  const colls = collsArg.split(",").map((s) => s.trim()).filter(Boolean);
  if (colls.length === 0) {
    throw new CommandError(EXIT.USAGE, "search-across needs at least one collection");
  }
  const entries = colls.map((c) => ({ coll: c, entry: requireVecColl(reg, c) }));
  const query = parseVectorJson(vecArg, ctx);
  const k = Number(flagString(parsed.flags, "k") ?? "10");
  const merged = [];
  for (const { coll: c, entry } of entries) {
    const hits = entry.store.search(c, query, k, 0, entry.metric, null);
    for (const h of hits) {
      merged.push({ id: h.id, score: h.score, coll: c, metadata: h.metadata });
    }
  }
  merged.sort((a, b) => b.score - a.score);
  return { stdout: JSON.stringify(merged.slice(0, k)) };
};
var getOp = async (reg, _ctx, parsed) => {
  const coll = parsed.positional[1];
  const id = parsed.positional[2];
  if (!coll || !id) throw new CommandError(EXIT.USAGE, "usage: vec get <coll> <id>");
  const entry = requireVecColl(reg, coll);
  const rec = entry.store.get(coll, id);
  if (!rec) throw new CommandError(EXIT.NOT_FOUND, `not found: ${coll}/${id}`);
  return { stdout: JSON.stringify({ id, ...rec }) };
};
var removeOp = async (reg, _ctx, parsed) => {
  const coll = parsed.positional[1];
  const id = parsed.positional[2];
  if (!coll || !id) throw new CommandError(EXIT.USAGE, "usage: vec remove <coll> <id>");
  const entry = requireVecColl(reg, coll);
  if (!entry.store.has(coll, id)) {
    throw new CommandError(EXIT.NOT_FOUND, `not found: ${coll}/${id}`);
  }
  entry.store.remove(coll, id);
  return { stdout: JSON.stringify({ removed: id }) };
};
var statsOp = async (reg, _ctx, parsed) => {
  const coll = parsed.positional[1];
  if (!coll) throw new CommandError(EXIT.USAGE, "usage: vec stats <coll>");
  const entry = requireVecColl(reg, coll);
  const suffix = entry.quantize === "int8" ? ".q8" : entry.quantize === "binary" ? ".b1" : entry.quantize === "polar" ? ".p3" : "";
  const bin = reg.mem.snapshotBin().get(`${coll}${suffix}.bin`);
  const meta = reg.mem.snapshotJson().get(`${coll}${suffix}.json`);
  const binBytes = bin?.byteLength ?? 0;
  const metaBytes = meta ? new TextEncoder().encode(JSON.stringify(meta)).byteLength : 0;
  const out = {
    dim: entry.dim,
    count: entry.store.count(coll),
    quantize: entry.quantize,
    metric: entry.metric,
    sizeBytes: binBytes + metaBytes,
    binBytes,
    metaBytes
  };
  if (reg.encBin && reg.encBin.isCorrupted(`${coll}${suffix}.bin`)) {
    out["corrupted"] = true;
  }
  if (entry.ivfIndex && entry.ivf) {
    out["ivf"] = {
      built: entry.ivfIndex.hasIndex(coll),
      numClusters: entry.ivf.numClusters,
      numProbes: entry.ivf.numProbes
    };
  }
  return { stdout: JSON.stringify(out) };
};
var dropOp = async (reg, _ctx, parsed) => {
  const coll = parsed.positional[1];
  if (!coll) throw new CommandError(EXIT.USAGE, "usage: vec drop <coll>");
  const entry = requireVecColl(reg, coll);
  entry.store.drop(coll);
  reg.removeVectorCollection(coll);
  reg.persistVectorRegistry();
  return { stdout: JSON.stringify({ dropped: coll }) };
};
var verifyOp = async (reg, _ctx, parsed) => {
  const coll = parsed.positional[1];
  if (!coll) throw new CommandError(EXIT.USAGE, "usage: vec verify <coll>");
  const entry = requireVecColl(reg, coll);
  const suffix = entry.quantize === "int8" ? ".q8" : entry.quantize === "binary" ? ".b1" : entry.quantize === "polar" ? ".p3" : "";
  const binFile = `${coll}${suffix}.bin`;
  const encrypted = reg.encBin !== null;
  const ok = !encrypted || !reg.encBin.isCorrupted(binFile);
  const out = {
    coll,
    ok,
    encrypted,
    binFile
  };
  if (!ok) {
    out["reason"] = "decrypt failed (wrong key, tampered ciphertext, or truncated IV)";
  }
  return { stdout: JSON.stringify(out) };
};
var exportOp = async (reg, _ctx, parsed) => {
  const coll = parsed.positional[1];
  if (!coll) throw new CommandError(EXIT.USAGE, "usage: vec export <coll>");
  const entry = requireVecColl(reg, coll);
  const records = entry.store.export(coll);
  return { stdout: JSON.stringify({ exported: records.length, records }) };
};
var importOp = async (reg, ctx, parsed) => {
  const coll = parsed.positional[1];
  const src = parsed.positional[2];
  if (!coll || !src) {
    throw new CommandError(EXIT.USAGE, "usage: vec import <coll> <jsonl-path-or-->");
  }
  const entry = requireVecColl(reg, coll);
  let text;
  if (src === "-") {
    text = ctx.stdin;
  } else {
    try {
      text = await ctx.fs.readFile(src, "utf8");
    } catch {
      throw new CommandError(EXIT.USAGE, `cannot read input: ${src}`);
    }
  }
  let raw;
  try {
    raw = JSON.parse(text);
  } catch {
    throw new CommandError(EXIT.VALIDATION, "validation: invalid JSON for import");
  }
  let parsedInput;
  if (Array.isArray(raw)) {
    parsedInput = raw;
  } else if (raw && typeof raw === "object" && Array.isArray(raw["records"])) {
    parsedInput = raw["records"];
  } else {
    throw new CommandError(EXIT.VALIDATION, "validation: import expects an array of records");
  }
  for (let i = 0; i < parsedInput.length; i++) {
    const check = checkVectorRecord(parsedInput[i], entry.dim);
    if (!check.ok) {
      throw new CommandError(
        EXIT.VALIDATION,
        `validation: import record at index ${i}: ${check.reason}`
      );
    }
  }
  entry.store.import(
    coll,
    parsedInput
  );
  return { stdout: JSON.stringify({ imported: parsedInput.length }) };
};

// src/commands/vec.ts
var FLAG_SPEC2 = {
  bool: ["json", "ivf", "no-ivf"],
  string: [
    "dim",
    "quantize",
    "metric",
    "k",
    "probe",
    "matryoshka",
    "meta",
    "ivf-clusters",
    "ivf-probes",
    "sample-dims"
  ]
};
var READ_ONLY = /* @__PURE__ */ new Set([
  "search",
  "search-across",
  "get",
  "stats",
  "export",
  "verify"
]);
var buildVecCommand = (provide) => justBash.defineCommand(
  "vec",
  async (args, ctx) => runCommand(async () => {
    const reg = provide(ctx.fs);
    const parsed = parseCommandArgs(args, FLAG_SPEC2);
    const sub = parsed.positional[0];
    if (!sub) {
      throw new CommandError(EXIT.USAGE, "usage: vec <subcommand> [...]");
    }
    await reg.ensureHydrated();
    let result;
    switch (sub) {
      case "create":
        result = await createOp(reg, ctx, parsed);
        break;
      case "store":
        result = await storeOp(reg, ctx, parsed);
        break;
      case "store-batch":
        result = await storeBatchOp(reg, ctx, parsed);
        break;
      case "search":
        result = await searchOp(reg, ctx, parsed);
        break;
      case "search-across":
        result = await searchAcrossOp(reg, ctx, parsed);
        break;
      case "get":
        result = await getOp(reg, ctx, parsed);
        break;
      case "remove":
        result = await removeOp(reg, ctx, parsed);
        break;
      case "stats":
        result = await statsOp(reg, ctx, parsed);
        break;
      case "import":
        result = await importOp(reg, ctx, parsed);
        break;
      case "export":
        result = await exportOp(reg, ctx, parsed);
        break;
      case "drop":
        result = await dropOp(reg, ctx, parsed);
        break;
      case "ivf":
        result = await ivfOp(reg, ctx, parsed);
        break;
      case "verify":
        result = await verifyOp(reg, ctx, parsed);
        break;
      default:
        throw new CommandError(EXIT.USAGE, `unknown subcommand: vec ${sub}`);
    }
    const isIvfReadOnly = sub === "ivf" && parsed.positional[1] === "stats";
    if (!READ_ONLY.has(sub) && !isIvfReadOnly) {
      for (const [, entry] of reg.vectorCollections()) {
        entry.store.flush();
      }
      await reg.flushIfDirty();
    }
    return result;
  })
);
var COMMON_COLLECTIONS = [
  // Observed in the 8-model benchmark transcripts:
  "books",
  "users",
  "docs",
  "notes",
  "chunks",
  "concepts",
  "people",
  // Common DB-domain names:
  "items",
  "events",
  "logs",
  "messages",
  "posts",
  "products",
  "orders",
  "comments",
  "sessions",
  "tokens",
  "files",
  "tasks",
  "jobs",
  "metrics",
  // Generic / placeholder / test idioms:
  "x",
  "y",
  "test",
  "demo",
  "coll",
  "collection",
  "data",
  "t",
  "foo",
  "bar"
];
var buildRedirectMessage = (tool, coll) => {
  const example = tool === "db" ? `${tool} ${coll} find '{}'` : `${tool} ${coll} stats`;
  return `'${tool}.${coll}' is MongoDB-shell-style syntax. This plugin uses space-separated form: '${tool} ${coll} <subcommand>'.
Example: ${example}
(Tip: the parenthesised form '${tool}.${coll}.method(...)' is rejected by bash itself before reaching the plugin \u2014 there is no way to intercept it from inside.)
`;
};
var buildSentinel = (tool, coll) => justBash.defineCommand(`${tool}.${coll}`, async () => ({
  stdout: "",
  stderr: buildRedirectMessage(tool, coll),
  exitCode: 2
}));
var buildSentinelCommands = () => {
  const all = [];
  for (const coll of COMMON_COLLECTIONS) {
    all.push(buildSentinel("db", coll));
    all.push(buildSentinel("vec", coll));
  }
  return all;
};
var sentinelNames = () => {
  const names = [];
  for (const coll of COMMON_COLLECTIONS) {
    names.push(`db.${coll}`);
    names.push(`vec.${coll}`);
  }
  return names;
};

// src/adapter.ts
var toUint8 = (data) => {
  if (data instanceof Uint8Array) return data;
  return new Uint8Array(data);
};
var toArrayBuffer = (u8) => u8.buffer.slice(u8.byteOffset, u8.byteOffset + u8.byteLength);
var MemoryAdapter = class {
  jsonStore = /* @__PURE__ */ new Map();
  binStore = /* @__PURE__ */ new Map();
  jsonDirty = /* @__PURE__ */ new Set();
  binDirty = /* @__PURE__ */ new Set();
  deletedSet = /* @__PURE__ */ new Set();
  readJson(name) {
    return this.jsonStore.has(name) ? this.jsonStore.get(name) : null;
  }
  writeJson(name, data) {
    this.jsonStore.set(name, data);
    this.jsonDirty.add(name);
    this.deletedSet.delete(name);
  }
  readBin(name) {
    const u8 = this.binStore.get(name);
    return u8 ? toArrayBuffer(u8) : null;
  }
  writeBin(name, data) {
    this.binStore.set(name, toUint8(data));
    this.binDirty.add(name);
    this.deletedSet.delete(name);
  }
  delete(name) {
    const had = this.jsonStore.delete(name) || this.binStore.delete(name);
    this.jsonDirty.delete(name);
    this.binDirty.delete(name);
    if (had) this.deletedSet.add(name);
  }
  loadJson(name, data) {
    this.jsonStore.set(name, data);
  }
  loadBin(name, data) {
    this.binStore.set(name, toUint8(data));
  }
  snapshotJson() {
    return this.jsonStore;
  }
  snapshotBin() {
    return this.binStore;
  }
  takeDirty() {
    const out = {
      jsonChanged: this.jsonDirty,
      binChanged: this.binDirty,
      deleted: this.deletedSet
    };
    this.jsonDirty = /* @__PURE__ */ new Set();
    this.binDirty = /* @__PURE__ */ new Set();
    this.deletedSet = /* @__PURE__ */ new Set();
    return out;
  }
  /**
   * Re-mark entries as dirty after a partial flush failure. The persister
   * calls this with the entries that did NOT successfully reach disk, so
   * the next flush retries them. Without this, takeDirty() on a partial
   * failure would silently drop pending writes.
   */
  restoreDirty(d) {
    if (d.jsonChanged) for (const n of d.jsonChanged) this.jsonDirty.add(n);
    if (d.binChanged) for (const n of d.binChanged) this.binDirty.add(n);
    if (d.deleted) for (const n of d.deleted) this.deletedSet.add(n);
  }
  hasDirty() {
    return this.jsonDirty.size > 0 || this.binDirty.size > 0 || this.deletedSet.size > 0;
  }
};
var getCrypto = () => {
  if (typeof globalThis.crypto !== "undefined" && globalThis.crypto.subtle) {
    return globalThis.crypto;
  }
  throw new Error("EncryptedBinAdapter: Web Crypto API not available");
};
var EncryptedBinAdapter = class _EncryptedBinAdapter {
  inner;
  key;
  cache = /* @__PURE__ */ new Map();
  pending = /* @__PURE__ */ new Set();
  /**
   * Names that failed to decrypt during preload. Surfaces a real failure
   * mode that would otherwise be invisible: a wrong key or corrupt ciphertext
   * caches as an empty buffer (so no plaintext leaks), but the user / agent
   * has no way to distinguish "empty collection" from "decrypt failed".
   * `vec stats` reads this Set to report `corrupted: true`.
   */
  corruptedSet = /* @__PURE__ */ new Set();
  constructor(inner, key) {
    this.inner = inner;
    this.key = key;
  }
  isCorrupted(name) {
    return this.corruptedSet.has(name);
  }
  static async create(inner, password, salt = "js-vector-store-v1") {
    const c = getCrypto();
    const enc = new TextEncoder();
    const keyMaterial = await c.subtle.importKey(
      "raw",
      enc.encode(password),
      "PBKDF2",
      false,
      ["deriveKey"]
    );
    const key = await c.subtle.deriveKey(
      {
        name: "PBKDF2",
        salt: enc.encode(salt),
        iterations: 1e5,
        hash: "SHA-256"
      },
      keyMaterial,
      { name: "AES-GCM", length: 256 },
      false,
      ["encrypt", "decrypt"]
    );
    return new _EncryptedBinAdapter(inner, key);
  }
  readJson(name) {
    return this.inner.readJson(name);
  }
  writeJson(name, data) {
    this.inner.writeJson(name, data);
  }
  readBin(name) {
    const u8 = this.cache.get(name);
    return u8 ? toArrayBuffer(u8) : null;
  }
  writeBin(name, data) {
    this.cache.set(name, toUint8(data));
    this.pending.add(name);
  }
  delete(name) {
    this.cache.delete(name);
    this.pending.delete(name);
    this.inner.delete(name);
  }
  async preload(names) {
    const c = getCrypto();
    for (const name of names) {
      const ctBuf = this.inner.readBin(name);
      if (!ctBuf || ctBuf.byteLength < 12) continue;
      try {
        const ivBuf = ctBuf.slice(0, 12);
        const bodyBuf = ctBuf.slice(12);
        const pt = await c.subtle.decrypt(
          { name: "AES-GCM", iv: ivBuf },
          this.key,
          bodyBuf
        );
        this.cache.set(name, new Uint8Array(pt));
      } catch {
        this.cache.set(name, new Uint8Array(0));
        this.corruptedSet.add(name);
      }
    }
  }
  async persist() {
    const c = getCrypto();
    for (const name of this.pending) {
      const pt = this.cache.get(name);
      if (!pt) continue;
      const iv = c.getRandomValues(new Uint8Array(12));
      const ptBuf = pt.buffer.slice(
        pt.byteOffset,
        pt.byteOffset + pt.byteLength
      );
      const ivBuf = iv.buffer.slice(
        iv.byteOffset,
        iv.byteOffset + iv.byteLength
      );
      const ct = await c.subtle.encrypt(
        { name: "AES-GCM", iv: ivBuf },
        this.key,
        ptBuf
      );
      const out = new Uint8Array(12 + ct.byteLength);
      out.set(iv, 0);
      out.set(new Uint8Array(ct), 12);
      this.inner.writeBin(name, out);
    }
    this.pending.clear();
  }
};

// src/persister.ts
var BIN_SUFFIX = ".bin";
var isBin = (name) => name.endsWith(BIN_SUFFIX);
var Persister = class {
  fs;
  root;
  dirEnsured = false;
  flushChain = Promise.resolve();
  constructor(fs, root) {
    this.fs = fs;
    this.root = root;
  }
  absPath(name) {
    return this.fs.resolvePath(this.root, name);
  }
  async ensureDir() {
    if (this.dirEnsured) return;
    await this.fs.mkdir(this.root, { recursive: true });
    this.dirEnsured = true;
  }
  async hydrate(mem, encJson, encBin) {
    await this.ensureDir();
    let entries;
    try {
      entries = await this.fs.readdir(this.root);
    } catch {
      return;
    }
    const tasks = entries.map(async (entry) => {
      if (entry.endsWith(".tmp")) return null;
      const abs = this.absPath(entry);
      const stat = await this.fs.stat(abs).catch(() => null);
      if (!stat || !stat.isFile) return null;
      if (isBin(entry)) {
        const buf = await this.fs.readFileBuffer(abs);
        return { kind: "bin", name: entry, data: buf };
      }
      if (entry.endsWith(".json")) {
        const text = await this.fs.readFile(abs, "utf8");
        try {
          return { kind: "json", name: entry, data: JSON.parse(text) };
        } catch {
          return null;
        }
      }
      return null;
    });
    const loaded = (await Promise.all(tasks)).filter((x) => x !== null);
    const jsonNames = [];
    const binNames = [];
    for (const item of loaded) {
      if (item.kind === "json") {
        mem.loadJson(item.name, item.data);
        jsonNames.push(item.name);
      } else {
        mem.loadBin(item.name, item.data);
        binNames.push(item.name);
      }
    }
    if (encJson && jsonNames.length > 0) await encJson.preload(jsonNames);
    if (encBin && binNames.length > 0) await encBin.preload(binNames);
  }
  async flush(mem, encJson, encBin) {
    const chained = this.flushChain.then(
      () => this.doFlush(mem, encJson ?? null, encBin ?? null)
    );
    this.flushChain = chained.catch(() => void 0);
    return chained;
  }
  async doFlush(mem, encJson, encBin) {
    if (encJson) await encJson.persist();
    if (encBin) await encBin.persist();
    const dirty = mem.takeDirty();
    if (dirty.jsonChanged.size === 0 && dirty.binChanged.size === 0 && dirty.deleted.size === 0) {
      return;
    }
    await this.ensureDir();
    const jsonSnap = mem.snapshotJson();
    const binSnap = mem.snapshotBin();
    const remainingJson = new Set(dirty.jsonChanged);
    const remainingBin = new Set(dirty.binChanged);
    const remainingDeleted = new Set(dirty.deleted);
    try {
      for (const name of dirty.jsonChanged) {
        const data = jsonSnap.get(name);
        if (data !== void 0) {
          await this.atomicWrite(name, JSON.stringify(data), "utf8");
        }
        remainingJson.delete(name);
      }
      for (const name of dirty.binChanged) {
        const data = binSnap.get(name);
        if (data) {
          await this.atomicWrite(name, data);
        }
        remainingBin.delete(name);
      }
      for (const name of dirty.deleted) {
        await this.fs.rm(this.absPath(name), { force: true });
        remainingDeleted.delete(name);
      }
    } catch (err) {
      mem.restoreDirty({
        jsonChanged: remainingJson,
        binChanged: remainingBin,
        deleted: remainingDeleted
      });
      throw err;
    }
  }
  async atomicWrite(name, payload, encoding) {
    const final = this.absPath(name);
    const tmp = this.absPath(`${name}.tmp`);
    try {
      if (encoding) {
        await this.fs.writeFile(tmp, payload, encoding);
      } else {
        await this.fs.writeFile(tmp, payload);
      }
      await this.fs.mv(tmp, final);
    } catch (err) {
      await this.fs.rm(tmp, { force: true }).catch(() => {
      });
      throw err;
    }
  }
};

// src/registry.ts
var VEC_REGISTRY_FILE = "_vec.registry.json";
var PluginRegistry = class {
  mem;
  opts;
  persister;
  root;
  encJson = null;
  encBin = null;
  docStoreInst = null;
  authInst = null;
  vecMap = /* @__PURE__ */ new Map();
  hydration = null;
  encReady = null;
  constructor(fs, opts) {
    this.mem = new MemoryAdapter();
    this.opts = opts;
    this.root = opts.rootDir ?? "/data";
    this.persister = new Persister(fs, this.root);
  }
  async setupEncryption() {
    if (this.encReady) return this.encReady;
    if (!this.opts.encryptionKey) return;
    const userSalt = this.opts.salt;
    this.encReady = (async () => {
      this.encJson = userSalt !== void 0 ? await jsDocStore.EncryptedAdapter.create(
        this.mem,
        this.opts.encryptionKey,
        `${userSalt}:json`
      ) : await jsDocStore.EncryptedAdapter.create(
        this.mem,
        this.opts.encryptionKey
      );
      this.encBin = userSalt !== void 0 ? await EncryptedBinAdapter.create(
        this.mem,
        this.opts.encryptionKey,
        `${userSalt}:bin`
      ) : await EncryptedBinAdapter.create(
        this.mem,
        this.opts.encryptionKey
      );
    })();
    return this.encReady;
  }
  async ensureHydrated() {
    if (this.hydration) return this.hydration;
    this.hydration = (async () => {
      await this.setupEncryption();
      await this.persister.hydrate(this.mem, this.encJson, this.encBin);
      this.rehydrateVectorRegistry();
    })();
    return this.hydration;
  }
  rehydrateVectorRegistry() {
    const raw = this.mem.readJson(VEC_REGISTRY_FILE);
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) return;
    for (const [coll, meta] of Object.entries(raw)) {
      if (!meta || typeof meta !== "object") continue;
      const m = meta;
      if (typeof m["dim"] !== "number" || typeof m["quantize"] !== "string" || typeof m["metric"] !== "string" || this.vecMap.has(coll)) {
        continue;
      }
      let ivf;
      const rawIvf = m["ivf"];
      if (rawIvf && typeof rawIvf === "object" && typeof rawIvf.numClusters === "number" && typeof rawIvf.numProbes === "number") {
        ivf = {
          numClusters: rawIvf.numClusters,
          numProbes: rawIvf.numProbes
        };
      }
      this.registerVectorCollection(
        coll,
        m["dim"],
        m["quantize"],
        m["metric"],
        ivf
      );
    }
  }
  persistVectorRegistry() {
    const out = {};
    for (const [coll, entry] of this.vecMap) {
      const o = {
        dim: entry.dim,
        quantize: entry.quantize,
        metric: entry.metric
      };
      if (entry.ivf) o.ivf = entry.ivf;
      out[coll] = o;
    }
    this.mem.writeJson(VEC_REGISTRY_FILE, out);
  }
  async flushIfDirty() {
    const hasEncPending = Boolean(this.encJson) || Boolean(this.encBin);
    if (!this.mem.hasDirty() && !hasEncPending) return;
    await this.persister.flush(this.mem, this.encJson, this.encBin);
  }
  getDocStore() {
    if (!this.docStoreInst) {
      const adapter = this.encJson ?? this.mem;
      this.docStoreInst = new jsDocStore.DocStore(adapter);
    }
    return this.docStoreInst;
  }
  async getAuth() {
    if (!this.opts.authSecret) {
      throw new Error("Auth not configured: pass authSecret in PluginOptions");
    }
    if (!this.authInst) {
      this.authInst = new jsDocStore.Auth(this.getDocStore(), {
        secret: this.opts.authSecret
      });
      await this.authInst.init();
    }
    return this.authInst;
  }
  getVectorCollection(coll) {
    return this.vecMap.get(coll) ?? null;
  }
  registerVectorCollection(coll, dim, quantize, metric, ivf) {
    if (this.vecMap.has(coll)) {
      throw new Error(`vector collection already exists: ${coll}`);
    }
    const adapter = this.encBin ?? this.mem;
    let store;
    switch (quantize) {
      case "int8":
        store = new jsVectorStore.QuantizedStore(adapter, dim);
        break;
      case "polar":
        store = new jsVectorStore.PolarQuantizedStore(adapter, dim);
        break;
      case "binary":
        store = new jsVectorStore.BinaryQuantizedStore(adapter, dim);
        break;
      default:
        store = new jsVectorStore.VectorStore(adapter, dim);
    }
    const entry = { store, dim, quantize, metric };
    if (ivf) {
      entry.ivf = ivf;
      entry.ivfIndex = new jsVectorStore.IVFIndex(store, ivf.numClusters, ivf.numProbes);
    }
    this.vecMap.set(coll, entry);
    return entry;
  }
  removeVectorCollection(coll) {
    return this.vecMap.delete(coll);
  }
  vectorCollections() {
    return this.vecMap.entries();
  }
};

// src/index.ts
var createDataPlugin = (opts = {}) => {
  const cache = /* @__PURE__ */ new WeakMap();
  const provide = (fs) => {
    let reg = cache.get(fs);
    if (!reg) {
      reg = new PluginRegistry(fs, opts);
      cache.set(fs, reg);
    }
    return reg;
  };
  return [
    buildDbCommand(provide),
    buildVecCommand(provide),
    ...buildSentinelCommands()
  ];
};

exports.createDataPlugin = createDataPlugin;
exports.sentinelNames = sentinelNames;
